tests/test-update-direct.c

```diff
@@ -0,0 +1,255 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Joaquim Rocha <jrocha@endlessm.com>
+ */
+
+#include <test-common/misc-utils.h>
+#include <test-common/spawn-utils.h>
+#include <test-common/utils.h>
+
+#include <gio/gio.h>
+#include <locale.h>
+
+#include "eos-updater-generated.h"
+
+typedef enum {
+	EOS_UPDATER_STATE_NONE = 0,
+	EOS_UPDATER_STATE_READY,
+	EOS_UPDATER_STATE_ERROR,
+	EOS_UPDATER_STATE_POLLING,
+	EOS_UPDATER_STATE_UPDATE_AVAILABLE,
+	EOS_UPDATER_STATE_FETCHING,
+	EOS_UPDATER_STATE_UPDATE_READY,
+	EOS_UPDATER_STATE_APPLYING_UPDATE,
+	EOS_UPDATER_STATE_UPDATE_APPLIED,
+        EOS_UPDATER_N_STATES,
+} EosUpdaterState;
+
+gboolean cancelled_states[EOS_UPDATER_N_STATES] = { FALSE };
+const gchar *states_str[] = {"None",
+                             "Ready",
+                             "Error",
+                             "Polling",
+                             "UpdateAvailable",
+                             "Fetching",
+                             "UpdateReady",
+                             "ApplyingUpdate",
+                             "UpdateApplied"
+};
+
+guint cancelled_error_count = 0;
+guint cancel_calls_count = 0;
+
+static gboolean
+cancel_update (EosUpdater *updater)
+{
+  gboolean should_succeed;
+  EosUpdaterState state = eos_updater_get_state (updater);
+  g_autoptr(GError) error = NULL;
+
+  switch (state)
+    {
+      case EOS_UPDATER_STATE_NONE:
+      case EOS_UPDATER_STATE_ERROR:
+      case EOS_UPDATER_STATE_READY:
+      case EOS_UPDATER_STATE_UPDATE_AVAILABLE:
+      case EOS_UPDATER_STATE_UPDATE_READY:
+      case EOS_UPDATER_STATE_UPDATE_APPLIED:
+        should_succeed = FALSE;
+        break;
+      case EOS_UPDATER_STATE_POLLING:
+      case EOS_UPDATER_STATE_FETCHING:
+      case EOS_UPDATER_STATE_APPLYING_UPDATE:
+      case EOS_UPDATER_N_STATES: /* just supress the warning about covering all enum values... */
+      default:
+        should_succeed = TRUE;
+        break;
+    }
+
+  g_debug ("Trying to cancel state %s", states_str[state]);
+
+  cancelled_states[state] = TRUE;
+  eos_updater_call_cancel_sync (updater, NULL, &error);
+
+  if (should_succeed)
+    {
+      g_assert_no_error (error);
+      ++cancel_calls_count;
+      g_debug ("Cancelled state %s", states_str[state]);
+    }
+  else
+    {
+      g_assert (error != NULL);
+      g_debug ("Error cancelling %s: %s", states_str[state], error->message);
```

Comment:
If the assertion fails, this `g_debug()` won’t be reached. How about making it a `g_test_message()` and moving it just above the `g_assert()`?

Reply:
I moved it under the assert on purpose. If the error is `NULL`, then the debug message shouldn't be shown as it'd not be true.

Reply:
Hmm, catch 22. Stick it in an `if` then, I guess.

Reply:
Why is it catch 22? If the error is NULL, then the assert will get it, in which case it makes no sense to print the debug msg `"Error cancelling"` because there's been no error...

Reply:
Sorry, brain fart. Ignore my comments here.